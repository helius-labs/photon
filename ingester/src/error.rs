use thiserror::Error;

#[derive(Error, Debug, PartialEq, Eq)]
pub enum IngesterError {
    #[error("Persist logic for {event_type} has not yet been implemented")]
    EventNotImplemented { event_type: String },
    #[error("Malformed event: {msg}")]
    MalformedEvent { msg: String },
    #[error("Database error: {0}")]
    DatabaseError(String),
    #[error("Ingestion error: {0}")]
    ParserError(String),
}

impl From<sea_orm::error::DbErr> for IngesterError {
    fn from(err: sea_orm::error::DbErr) -> Self {
        IngesterError::DatabaseError(format!("DatabaseError: {}", err.to_string()))
    }
}
