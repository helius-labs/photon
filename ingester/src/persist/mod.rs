use crate::{
    error,
    parser::bundle::{ChangelogEvent, EventBundle, PublicStateTransitionBundle, UTXOEvent},
};
use dao::generated::{state_trees, utxos};
use log::{debug, warn};
use sea_orm::{
    sea_query::OnConflict, ColumnTrait, ConnectionTrait, DatabaseConnection, DatabaseTransaction,
    DbBackend, DbErr, EntityTrait, QueryFilter, QueryTrait, Set, TransactionTrait,
};
use solana_sdk::signature::Signature;

use error::IngesterError;

pub async fn persist_bundle(
    conn: &DatabaseConnection,
    event: EventBundle,
) -> Result<(), IngesterError> {
    match event {
        EventBundle::PublicStateTransition(e) => {
            debug!("Persisting PublicStateTransitionBundle: {:?}", e);
            persist_state_transition(conn, e).await
        }
        EventBundle::PublicNullifier(e) => {
            debug!("Persisting PublicNullifierBundle: {:?}", e);
            Err(IngesterError::EventNotImplemented {
                event_type: "PublicNullifierBundle".to_string(),
            })
        }
    }
}

async fn persist_state_transition(
    conn: &DatabaseConnection,
    bundle: PublicStateTransitionBundle,
) -> Result<(), IngesterError> {
    let PublicStateTransitionBundle {
        in_utxos,
        out_utxos,
        changelogs,
        transaction,
        slot_updated,
    } = bundle;

    if out_utxos.len() != changelogs.len() {
        return Err(IngesterError::MalformedEvent {
            msg: format!(
                "Number of changelogs did not match the number of output UTXOs (txn: {})",
                transaction,
            ),
        });
    }

    let txn = conn.begin().await?;
    for in_utxo in in_utxos {
        spend_input_utxo(&txn, &in_utxo, slot_updated, transaction).await?;
    }
    for i in 0..out_utxos.len() {
        let out_utxo = &out_utxos[i];
        let cl_event = &changelogs[i];
        append_output_utxo(&txn, out_utxo, slot_updated).await?;
        persist_changelog_event(&txn, cl_event, slot_updated).await?;
    }
    txn.commit().await?;
    Ok(())
}

async fn spend_input_utxo(
    txn: &DatabaseTransaction,
    in_utxo: &UTXOEvent,
    slot_updated: i64,
    transaction: Signature,
) -> Result<(), IngesterError> {
    let UTXOEvent { hash, seq, .. } = in_utxo;
    let model = utxos::ActiveModel {
        data: Set(vec![]),
        spent: Set(true),
        owner: Set(vec![]),
        slot_updated: Set(slot_updated),
        seq: Set(*seq),
        ..Default::default()
    };

    // Optimistically update the UTXO.
    let res = utxos::Entity::update(model)
        .filter(utxos::Column::Hash.eq(hash.to_vec()))
        .exec(txn)
        .await;

    // If data arrived out of order, the UTXO may not yet exist.
    // In this case we insert a placeholder marking the UTXO as spent.
    if let Err(e) = res {
        match e {
            DbErr::RecordNotFound(_) => {
                warn!("Input UTXO not found (hash: {}, txn: {}", hash, transaction);
                utxos::Entity::insert(utxos::ActiveModel {
                    hash: Set(hash.to_vec()),
                    data: Set(vec![]),
                    owner: Set(vec![]),
                    lamports: Set(None),
                    spent: Set(true),
                    slot_updated: Set(slot_updated),
                    seq: Set(*seq),
                    ..Default::default()
                })
                .on_conflict(
                    OnConflict::column(utxos::Column::Hash)
                        .update_columns([
                            utxos::Column::Hash,
                            utxos::Column::Data,
                            utxos::Column::Lamports,
                            utxos::Column::Spent,
                            utxos::Column::Seq,
                            utxos::Column::SlotUpdated,
                        ])
                        .to_owned(),
                )
                .exec(txn)
                .await?;
            }
            _ => return Err(e.into()),
        }
    }
    Ok(())
}

async fn append_output_utxo(
    txn: &DatabaseTransaction,
    out_utxo: &UTXOEvent,
    slot_updated: i64,
) -> Result<(), IngesterError> {
    let model = utxos::ActiveModel {
        hash: Set(out_utxo.hash.to_vec()),
        data: Set(out_utxo.data.clone()),
        tree: Set(out_utxo.tree.as_ref().to_vec()),
        account: Set(out_utxo.account.map(|a| a.as_ref().to_vec())),
        owner: Set(out_utxo.owner.as_ref().to_vec()),
        lamports: Set(out_utxo.lamports),
        spent: Set(false),
        seq: Set(out_utxo.seq),
        slot_updated: Set(slot_updated),
        ..Default::default()
    };

    // The state tree is append-only for output UTXOs.
    // Therefore we never anticpate conflicts unless the record is already inserted
    // or has been marked as spent.
    let stmt = utxos::Entity::insert(model)
        .on_conflict(
            OnConflict::column(utxos::Column::Hash)
                .do_nothing()
                .to_owned(),
        )
        .build(DbBackend::Postgres);
    txn.execute(stmt).await?;

    Ok(())
}

async fn persist_changelog_event(
    txn: &DatabaseTransaction,
    event: &ChangelogEvent,
    slot_updated: i64,
) -> Result<(), IngesterError> {
    let ChangelogEvent { path, tree, seq } = event;
    let depth = path.len() - 1;
    let items = path
        .iter()
        .enumerate()
        .map(|(i, p)| {
            let node_idx = p.index as i64;
            let leaf_idx = if i == 0 {
                Some(node_idx_to_leaf_idx(node_idx, depth as u32))
            } else {
                None
            };
            state_trees::ActiveModel {
                tree: Set(tree.as_ref().to_vec()),
                level: Set(i as i64),
                node_idx: Set(node_idx),
                hash: Set(p.hash.to_vec()),
                leaf_idx: Set(leaf_idx),
                seq: Set(*seq as i64),
                slot_updated: Set(slot_updated),
                ..Default::default()
            }
        })
        .collect::<Vec<_>>();

    let mut query = state_trees::Entity::insert_many(items)
        .on_conflict(
            OnConflict::columns([state_trees::Column::Tree, state_trees::Column::NodeIdx])
                .update_columns([
                    state_trees::Column::Hash,
                    state_trees::Column::Seq,
                    state_trees::Column::SlotUpdated,
                ])
                .to_owned(),
        )
        .build(DbBackend::Postgres);
    query.sql = format!("{} WHERE excluded.seq > state_trees.seq", query.sql);
    txn.execute(query).await?;

    Ok(())
}

fn node_idx_to_leaf_idx(index: i64, tree_height: u32) -> i64 {
    index - 2i64.pow(tree_height)
}
