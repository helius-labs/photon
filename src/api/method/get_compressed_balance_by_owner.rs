use crate::common::typedefs::serializable_pubkey::SerializablePubkey;
use crate::common::typedefs::unsigned_integer::UnsignedInteger;
use crate::dao::generated::owner_balances;
use sea_orm::{ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QuerySelect};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::super::error::PhotonApiError;
use super::utils::Context;
use super::utils::{parse_decimal, AccountBalanceResponse, LamportModel};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, ToSchema, Default)]
#[serde(rename_all = "camelCase")]
pub struct GetCompressedBalanceByOwnerRequest {
    pub owner: SerializablePubkey,
}

pub async fn get_compressed_balance_by_owner(
    conn: &DatabaseConnection,
    request: GetCompressedBalanceByOwnerRequest,
) -> Result<AccountBalanceResponse, PhotonApiError> {
    let context = Context::extract(conn).await?;
    let owner = request.owner;

    let balances = owner_balances::Entity::find()
        .select_only()
        .column(owner_balances::Column::Lamports)
        .filter(owner_balances::Column::Owner.eq::<Vec<u8>>(owner.into()))
        .into_model::<LamportModel>()
        .all(conn)
        .await?
        .iter()
        .map(|x| parse_decimal(x.lamports))
        .collect::<Result<Vec<u64>, PhotonApiError>>()?;

    let total_balance = balances.iter().sum::<u64>();
    if total_balance == 0 {
        return Err(PhotonApiError::RecordNotFound(format!(
            "No balance found for owner: {}",
            owner
        )));
    }

    Ok(AccountBalanceResponse {
        value: UnsignedInteger(total_balance),
        context,
    })
}
