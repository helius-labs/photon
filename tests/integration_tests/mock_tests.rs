use crate::utils::*;
use ::borsh::{to_vec, BorshDeserialize, BorshSerialize};
use function_name::named;
use photon_indexer::api::error::PhotonApiError;
use photon_indexer::api::method::get_compressed_accounts_by_owner::GetCompressedAccountsByOwnerRequest;
use photon_indexer::api::method::get_compressed_balance_by_owner::GetCompressedBalanceByOwnerRequest;
use photon_indexer::api::method::get_compressed_token_balances_by_owner::GetCompressedTokenBalancesByOwner;
use photon_indexer::api::method::get_multiple_compressed_accounts::GetMultipleCompressedAccountsRequest;
use photon_indexer::api::method::utils::{
    CompressedAccountRequest, GetCompressedTokenAccountsByDelegate,
    GetCompressedTokenAccountsByOwner,
};
use photon_indexer::common::typedefs::account::Account;
use photon_indexer::common::typedefs::bs64_string::Base64String;
use photon_indexer::common::typedefs::{hash::Hash, serializable_pubkey::SerializablePubkey};
use photon_indexer::dao::generated::accounts;
use photon_indexer::ingester::index_block;
use photon_indexer::ingester::parser::state_update::StateUpdate;
use photon_indexer::ingester::persist::{
    persist_state_update, persist_token_accounts, EnrichedTokenAccount,
};
use photon_indexer::ingester::typedefs::block_info::{BlockInfo, BlockMetadata};
use sea_orm::{EntityTrait, Set};
use serial_test::serial;

use photon_indexer::{
    common::typedefs::account::AccountData,
    ingester::parser::{indexer_events::PathNode, state_update::EnrichedPathNode},
};
use std::collections::HashMap;

use photon_indexer::common::typedefs::token_data::{AccountState, TokenData};
use sqlx::types::Decimal;

use photon_indexer::api::method::utils::Limit;
use solana_sdk::pubkey::Pubkey;
use std::vec;

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
struct Person {
    name: String,
    age: u64,
}

// TODO:
// - Add tests for duplicate inserts.
// - Add tests for accounts input spends without existing accounts.
// - Add test for multi-input/output transitions.

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_persist_state_update_basic(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup(name, db_backend).await;

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    let mut state_update = StateUpdate::new();
    let account = Account {
        hash: Hash::new_unique(),
        address: Some(SerializablePubkey::new_unique()),
        data: Some(AccountData {
            discriminator: 1,
            data: Base64String(vec![1; 500]),
            data_hash: Hash::new_unique(),
        }),
        owner: SerializablePubkey::new_unique(),
        lamports: 1000,
        tree: SerializablePubkey::new_unique(),
        leaf_index: 0,
        seq: Some(0),
        slot_updated: 0,
    };

    state_update.out_accounts.push(account.clone());
    persist_state_update_using_connection(&setup.db_conn, state_update)
        .await
        .unwrap();

    let request = CompressedAccountRequest {
        address: None,
        hash: Some(Hash::from(account.hash.clone())),
    };

    let res = setup
        .api
        .get_compressed_account(request.clone())
        .await
        .unwrap()
        .value;

    assert_eq!(res, account);

    let res = setup
        .api
        .get_compressed_balance(request)
        .await
        .unwrap()
        .value;

    assert_eq!(res, account.lamports as u64);

    // Assert that we get an error if we input a non-existent account.
    // TODO: Test spent accounts
    let err = setup
        .api
        .get_compressed_account(CompressedAccountRequest {
            hash: Some(Hash::from(Pubkey::new_unique().to_bytes())),
            address: None,
        })
        .await
        .unwrap_err();

    match err {
        PhotonApiError::RecordNotFound(_) => {}
        _ => panic!("Expected NotFound error"),
    }
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_multiple_accounts(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup(name, db_backend).await;

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    let owner1 = SerializablePubkey::new_unique();
    let owner2 = SerializablePubkey::new_unique();
    let mut state_update = StateUpdate::default();

    let accounts = vec![
        Account {
            hash: Hash::new_unique(),
            address: Some(SerializablePubkey::new_unique()),
            data: Some(AccountData {
                discriminator: 0,
                data: Base64String(vec![1; 500]),
                data_hash: Hash::new_unique(),
            }),
            owner: owner1,
            lamports: 1000,
            tree: SerializablePubkey::new_unique(),
            leaf_index: 10,
            seq: Some(1),
            slot_updated: 0,
        },
        Account {
            hash: Hash::new_unique(),
            address: None,
            data: Some(AccountData {
                discriminator: 1,
                data: Base64String(vec![2; 500]),
                data_hash: Hash::new_unique(),
            }),
            owner: owner1,
            lamports: 1030,
            tree: SerializablePubkey::new_unique(),
            leaf_index: 11,
            seq: Some(2),
            slot_updated: 0,
        },
        Account {
            hash: Hash::new_unique(),
            address: Some(SerializablePubkey::new_unique()),
            data: Some(AccountData {
                discriminator: 4,
                data: Base64String(vec![4; 500]),
                data_hash: Hash::new_unique(),
            }),
            owner: owner2,
            lamports: 10020,
            tree: SerializablePubkey::new_unique(),
            leaf_index: 13,
            seq: Some(3),
            slot_updated: 1,
        },
        Account {
            hash: Hash::new_unique(),
            address: Some(SerializablePubkey::new_unique()),
            data: Some(AccountData {
                discriminator: 10,
                data: Base64String(vec![5; 500]),
                data_hash: Hash::new_unique(),
            }),
            owner: owner2,
            lamports: 10100,
            tree: SerializablePubkey::new_unique(),
            leaf_index: 23,
            seq: Some(1),
            slot_updated: 0,
        },
    ];
    state_update.out_accounts = accounts.clone();
    persist_state_update_using_connection(&setup.db_conn, state_update)
        .await
        .unwrap();

    for owner in [owner1, owner2] {
        let res = setup
            .api
            .get_compressed_accounts_by_owner(GetCompressedAccountsByOwnerRequest {
                owner: SerializablePubkey::from(owner),
                ..Default::default()
            })
            .await
            .unwrap()
            .value;

        let mut response_accounts = res.items;

        let mut paginated_response_accounts = Vec::new();
        let mut cursor = None;
        loop {
            let res = setup
                .api
                .get_compressed_accounts_by_owner(GetCompressedAccountsByOwnerRequest {
                    owner: SerializablePubkey::from(owner),
                    cursor: cursor.clone(),
                    limit: Some(Limit::new(1).unwrap()),
                })
                .await
                .unwrap()
                .value;

            paginated_response_accounts.extend(res.items.clone());
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_eq!(response_accounts, paginated_response_accounts);

        let mut accounts_of_interest = accounts
            .clone()
            .into_iter()
            .filter(|x| x.owner == owner)
            .collect::<Vec<Account>>();

        assert_account_response_list_matches_input(
            &mut response_accounts,
            &mut accounts_of_interest,
        );

        let total_balance = accounts_of_interest
            .iter()
            .fold(0, |acc, x| acc + x.lamports);

        let res = setup
            .api
            .get_compressed_balance_by_owner(GetCompressedBalanceByOwnerRequest {
                owner: SerializablePubkey::from(owner),
            })
            .await
            .unwrap()
            .value;

        assert_eq!(res, total_balance);
    }

    let mut accounts_of_interest = vec![accounts[0].clone(), accounts[2].clone()];
    let mut res = setup
        .api
        .get_multiple_compressed_accounts(GetMultipleCompressedAccountsRequest {
            addresses: None,
            hashes: Some(
                accounts_of_interest
                    .iter()
                    .map(|x| x.hash.clone())
                    .collect(),
            ),
        })
        .await
        .unwrap()
        .value;

    assert_account_response_list_matches_input(&mut res.items, &mut accounts_of_interest);
}

#[named]
#[rstest]
#[tokio::test]
#[serial]
async fn test_persist_token_data(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    let name = trim_test_name(function_name!());
    let setup = setup(name, db_backend).await;
    let mint1 = SerializablePubkey::new_unique();
    let mint2 = SerializablePubkey::new_unique();
    let mint3 = SerializablePubkey::new_unique();
    let owner1 = SerializablePubkey::new_unique();
    let owner2 = SerializablePubkey::new_unique();
    let delegate1 = SerializablePubkey::new_unique();
    let delegate2 = SerializablePubkey::new_unique();

    // HACK: We index a block so that API methods can fetch the current slot.
    index_block(
        &setup.db_conn,
        &BlockInfo {
            metadata: BlockMetadata {
                slot: 0,
                ..Default::default()
            },
            ..Default::default()
        },
    )
    .await
    .unwrap();

    let token_data1 = TokenData {
        mint: mint1,
        owner: owner1,
        amount: 1,
        delegate: Some(delegate1),
        state: AccountState::frozen,
        is_native: Some(1),
        delegated_amount: 1,
    };

    let token_data2 = TokenData {
        mint: mint2,
        owner: owner1,
        amount: 2,
        delegate: Some(delegate2),
        state: AccountState::initialized,
        is_native: None,
        delegated_amount: 2,
    };

    let token_data3 = TokenData {
        mint: mint3,
        owner: owner2,
        amount: 3,
        delegate: Some(delegate1),
        state: AccountState::frozen,
        is_native: Some(1000),
        delegated_amount: 3,
    };
    let all_token_data = vec![token_data1, token_data2, token_data3];

    let txn = sea_orm::TransactionTrait::begin(setup.db_conn.as_ref())
        .await
        .unwrap();

    let mut token_datas = Vec::new();

    for (i, token_data) in all_token_data.iter().enumerate() {
        let slot = 11;
        let hash = Hash::new_unique();
        let model = accounts::ActiveModel {
            hash: Set(hash.clone().into()),
            address: Set(Some(Pubkey::new_unique().to_bytes().to_vec())),
            spent: Set(false),
            data: Set(Some(to_vec(&token_data).unwrap())),
            owner: Set(token_data.owner.to_bytes_vec()),
            lamports: Set(Decimal::from(10)),
            slot_updated: Set(slot),
            leaf_index: Set(i as i64),
            discriminator: Set(Some(Decimal::from(1))),
            data_hash: Set(Some(Hash::new_unique().to_vec())),
            tree: Set(Pubkey::new_unique().to_bytes().to_vec()),
            ..Default::default()
        };
        accounts::Entity::insert(model).exec(&txn).await.unwrap();
        token_datas.push(EnrichedTokenAccount {
            hash,
            token_data: token_data.clone(),
        });
    }

    persist_token_accounts(&txn, token_datas).await.unwrap();
    txn.commit().await.unwrap();

    let owner_tlv = all_token_data
        .iter()
        .filter(|x| x.owner == owner1 && x.mint == mint1)
        .map(Clone::clone)
        .collect();

    let res = setup
        .api
        .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
            owner: SerializablePubkey::from(owner1),
            mint: Some(SerializablePubkey::from(mint1)),
            ..Default::default()
        })
        .await
        .unwrap()
        .value;
    verify_responses_match_tlv_data(res.clone(), owner_tlv);

    for owner in [owner1, owner2] {
        let owner_tlv = all_token_data
            .iter()
            .filter(|x| x.owner == owner)
            .map(Clone::clone)
            .collect();
        let res = setup
            .api
            .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
                owner: SerializablePubkey::from(owner),
                ..Default::default()
            })
            .await
            .unwrap()
            .value;

        let mut paginated_res = Vec::new();
        let mut cursor = None;
        loop {
            let res = setup
                .api
                .get_compressed_token_accounts_by_owner(GetCompressedTokenAccountsByOwner {
                    owner: SerializablePubkey::from(owner),
                    cursor: cursor.clone(),
                    limit: Some(photon_indexer::api::method::utils::Limit::new(1).unwrap()),
                    ..Default::default()
                })
                .await
                .unwrap()
                .value;

            paginated_res.extend(res.items.clone());
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_eq!(paginated_res, res.items);

        let mut mint_to_balance: HashMap<SerializablePubkey, u64> = HashMap::new();

        for token_account in paginated_res.iter() {
            let balance = mint_to_balance
                .entry(token_account.token_data.mint.clone())
                .or_insert(0);
            *balance += token_account.token_data.amount;
        }
        for (mint, balance) in mint_to_balance.iter() {
            let request = GetCompressedTokenBalancesByOwner {
                owner: SerializablePubkey::from(owner),
                mint: Some(mint.clone()),
                ..Default::default()
            };
            let res = setup
                .api
                .get_compressed_token_balances_by_owner(request)
                .await
                .unwrap()
                .value;
            assert_eq!(res.token_balances[0].balance, *balance);
        }

        verify_responses_match_tlv_data(res.clone(), owner_tlv);
        for token_account in res.items {
            let request = CompressedAccountRequest {
                address: None,
                hash: Some(token_account.account.hash),
            };
            let balance = setup
                .api
                .get_compressed_token_account_balance(request)
                .await
                .unwrap()
                .value;
            assert_eq!(
                balance.amount,
                Into::<u64>::into(token_account.token_data.amount)
            );
        }
    }
    for delegate in [delegate1, delegate2] {
        let delegate_tlv = all_token_data
            .clone()
            .into_iter()
            .filter(|x| x.delegate == Some(delegate))
            .collect();
        let res = setup
            .api
            .get_compressed_token_accounts_by_delegate(GetCompressedTokenAccountsByDelegate {
                delegate: SerializablePubkey::from(delegate),
                ..Default::default()
            })
            .await
            .unwrap()
            .value;
        let mut paginated_res = Vec::new();
        let mut cursor = None;
        loop {
            let res = setup
                .api
                .get_compressed_token_accounts_by_delegate(GetCompressedTokenAccountsByDelegate {
                    delegate: SerializablePubkey::from(delegate),
                    cursor: cursor.clone(),
                    limit: Some(photon_indexer::api::method::utils::Limit::new(1).unwrap()),
                    ..Default::default()
                })
                .await
                .unwrap()
                .value;

            paginated_res.extend(res.items.clone());
            cursor = res.cursor;
            if cursor.is_none() {
                break;
            }
        }
        assert_eq!(paginated_res, res.items);
        verify_responses_match_tlv_data(res, delegate_tlv)
    }
}

#[named]
#[rstest]
#[tokio::test(flavor = "multi_thread", worker_threads = 25)]
#[serial]
#[ignore]
/// Test for testing how fast we can index accounts.
async fn test_load_test(
    #[values(DatabaseBackend::Sqlite, DatabaseBackend::Postgres)] db_backend: DatabaseBackend,
) {
    use std::collections::HashSet;

    let name = trim_test_name(function_name!());
    let setup = setup(name, db_backend).await;

    fn generate_random_account(_tree: Pubkey, _seq: i64) -> Account {
        Account {
            hash: Hash::new_unique(),
            address: Some(SerializablePubkey::new_unique()),
            data: Some(AccountData {
                discriminator: 10,
                data: Base64String(vec![5; 500]),
                data_hash: Hash::new_unique(),
            }),
            owner: SerializablePubkey::new_unique(),
            lamports: 10100,
            tree: SerializablePubkey::new_unique(),
            leaf_index: 23,
            seq: Some(1),
            slot_updated: 0,
        }
    }

    fn generate_random_leaf_index(tree: Pubkey, node_index: u32, seq: i64) -> EnrichedPathNode {
        EnrichedPathNode {
            node: PathNode {
                node: Pubkey::new_unique().to_bytes(),
                index: node_index,
            },
            slot: 0,
            tree: tree.to_bytes(),
            seq: seq as u64,
            level: 0,
            leaf_index: Some(seq as u32),
            tree_depth: 20,
        }
    }

    let loops = 25;
    for _ in 0..loops {
        let tree: Pubkey = Pubkey::new_unique();
        let txn = sea_orm::TransactionTrait::begin(setup.db_conn.as_ref())
            .await
            .unwrap();
        let num_elements = 2000;
        let state_update = StateUpdate {
            in_accounts: vec![],
            out_accounts: (0..num_elements)
                .map(|i| generate_random_account(tree, i))
                .collect(),
            // We only include the leaf index because we think the most path nodes will be
            // overwritten anyways. So the amortized number of writes will be in each tree
            // will be close to 1.
            path_nodes: (0..num_elements)
                .map(|i| {
                    let leaf = generate_random_leaf_index(tree, i as u32, i);
                    let key = (leaf.tree.clone(), leaf.leaf_index.unwrap());
                    (key, leaf)
                })
                .collect(),
            account_transactions: HashSet::new(),
        };
        persist_state_update(&txn, state_update).await.unwrap();
        txn.commit().await.unwrap();
    }
}
