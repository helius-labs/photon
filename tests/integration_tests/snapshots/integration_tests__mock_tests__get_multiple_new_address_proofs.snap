---
source: tests/integration_tests/mock_tests.rs
expression: proof
---
{
  "context": {
    "slot": 0
  },
  "value": [
    {
      "root": "3FrGzyXtjqjnukHDS2M5oyVj8tMVcHSahcHpWBGR5MHY",
      "address": "Fi6AXBGuGs7DRXP428hwhJJfTpJ4BVZD8DiUcX1cj35W",
      "lowerRangeAddress": "11111111111111111111111111111111",
      "higherRangeAddress": "14uQeVj5tqViQh7yWWGStvkEG1Zmhx6uasJtWCJziofL",
      "leafIndex": 1,
      "proof": [
        "34cMT7MjFrs8hLp2zHMrPJHKkUxBDBwBTNck77wLjjcY",
        "3CFKteYSRSkp9EGFPzibgzCPtNuJ5HeZ31fAsBpgcHwZ",
        "274k4KPY5VhmU5pCwwJzyHaY4AH4m7SjnZnyj44UitJc",
        "2gQksfK543Jw196amPxXWceFzNTfRHiqb7x2edvupftT",
        "Y8pRoGMGXJPGZyb7m4sqkPsTSJvzZ7oSVnaybCcYbgd",
        "3w895nemtSxb1fEjFT3cXkVpW3y6f9uHQn43jEwu1RWc",
        "46JM2e72mdEAxkSeXc65b3aLWkAQpX6NF7mdyuXQKguu",
        "WKLyYL7wpK5zVsWoMU2rcUiMwRWp9PowG83GHHH1dqn",
        "4CztDUcRvGkTyFoVEXXcBwd5UwqJcJJWpdL8FFERrqkU",
        "yjDkqNqZtBwpG7iUwT3hqKNKZAP9CcBDyBBuGPNk2jQ",
        "2r8qWibedWpM7QkW6QvDDUEQAHT9isX629UKcUxbY8BT",
        "38Aoz1ZUAnRqSCxdQuEkLhzYMRi1Tu1rSBy1UymA8yfK",
        "3zBeAgo5EwPrgsXRLUV6DhQrsUUhwbmQXPnZkudhsnSD",
        "2Q5cDvZW6hTxFFDTHhAf129PH4z8EZLYA1meeZpPW76F",
        "2gnsMw1zhpqGeAakoQLarCXVUaeEGDHm5uma1Jznx7qD",
        "3MXZLrn3tnLMWJznkMombRHkgkfsiBTPru2wjHDd3kLh",
        "3rrDm8iED4sDyonpLs9J3TDoTP8U4jsvXCmoupoCvvKc",
        "48YJtP1G5pKWu7Jy1NRagdj29dwxVXFsySrBVw1nALhX",
        "22tjwuiv6VgeMDs8qXpkZ7cY5rcgrFrG5knp7swZ1wQ2",
        "2dS4UCZ6RJvwPBaWmcz1L7ips34xnbNMfK9D8UHdFJfT",
        "3EdF4zvu7uwW7xYdnaKUPzqDY5Msq9FfWn6HgEnRTkDB",
        "2jzv6Z8vEXSYRXQCFnVYvzgJ2RfyDN1QEjdHm6CBvJUg",
        "2fomPS2xUPwXxVF8hx5YtMpBHmQKSZKushUiWYsy9Eqx",
        "3Pi3nRu4fWpnW5rYDAoraX1Q2xDCRQ24Dq8TEY7gaSBn",
        "3dbMMyehykAriNZ8FMJgPc1N6YSzm9ZfV2HcW9sbaXzP",
        "4EE5KQnQ6Tvo78gGhTu8serpJV4srbRhYZ3rdvCiRa5e"
      ],
      "merkleTree": "C83cpRN6oaafjNgMQJvaYgAz592EP5wunKvbokeTKPLn",
      "rootSeq": 3,
      "lowElementLeafIndex": 0
    },
    {
      "root": "3FrGzyXtjqjnukHDS2M5oyVj8tMVcHSahcHpWBGR5MHY",
      "address": "sH8ux4csv8wxiRejuHjpTCkrfGgeNtg6Y55AJJ9GJSd",
      "lowerRangeAddress": "11111111111111111111111111111111",
      "higherRangeAddress": "14uQeVj5tqViQh7yWWGStvkEG1Zmhx6uasJtWCJziofL",
      "leafIndex": 1,
      "proof": [
        "34cMT7MjFrs8hLp2zHMrPJHKkUxBDBwBTNck77wLjjcY",
        "3CFKteYSRSkp9EGFPzibgzCPtNuJ5HeZ31fAsBpgcHwZ",
        "274k4KPY5VhmU5pCwwJzyHaY4AH4m7SjnZnyj44UitJc",
        "2gQksfK543Jw196amPxXWceFzNTfRHiqb7x2edvupftT",
        "Y8pRoGMGXJPGZyb7m4sqkPsTSJvzZ7oSVnaybCcYbgd",
        "3w895nemtSxb1fEjFT3cXkVpW3y6f9uHQn43jEwu1RWc",
        "46JM2e72mdEAxkSeXc65b3aLWkAQpX6NF7mdyuXQKguu",
        "WKLyYL7wpK5zVsWoMU2rcUiMwRWp9PowG83GHHH1dqn",
        "4CztDUcRvGkTyFoVEXXcBwd5UwqJcJJWpdL8FFERrqkU",
        "yjDkqNqZtBwpG7iUwT3hqKNKZAP9CcBDyBBuGPNk2jQ",
        "2r8qWibedWpM7QkW6QvDDUEQAHT9isX629UKcUxbY8BT",
        "38Aoz1ZUAnRqSCxdQuEkLhzYMRi1Tu1rSBy1UymA8yfK",
        "3zBeAgo5EwPrgsXRLUV6DhQrsUUhwbmQXPnZkudhsnSD",
        "2Q5cDvZW6hTxFFDTHhAf129PH4z8EZLYA1meeZpPW76F",
        "2gnsMw1zhpqGeAakoQLarCXVUaeEGDHm5uma1Jznx7qD",
        "3MXZLrn3tnLMWJznkMombRHkgkfsiBTPru2wjHDd3kLh",
        "3rrDm8iED4sDyonpLs9J3TDoTP8U4jsvXCmoupoCvvKc",
        "48YJtP1G5pKWu7Jy1NRagdj29dwxVXFsySrBVw1nALhX",
        "22tjwuiv6VgeMDs8qXpkZ7cY5rcgrFrG5knp7swZ1wQ2",
        "2dS4UCZ6RJvwPBaWmcz1L7ips34xnbNMfK9D8UHdFJfT",
        "3EdF4zvu7uwW7xYdnaKUPzqDY5Msq9FfWn6HgEnRTkDB",
        "2jzv6Z8vEXSYRXQCFnVYvzgJ2RfyDN1QEjdHm6CBvJUg",
        "2fomPS2xUPwXxVF8hx5YtMpBHmQKSZKushUiWYsy9Eqx",
        "3Pi3nRu4fWpnW5rYDAoraX1Q2xDCRQ24Dq8TEY7gaSBn",
        "3dbMMyehykAriNZ8FMJgPc1N6YSzm9ZfV2HcW9sbaXzP",
        "4EE5KQnQ6Tvo78gGhTu8serpJV4srbRhYZ3rdvCiRa5e"
      ],
      "merkleTree": "C83cpRN6oaafjNgMQJvaYgAz592EP5wunKvbokeTKPLn",
      "rootSeq": 3,
      "lowElementLeafIndex": 0
    }
  ]
}
